Some tentative improvements to `Dispatcher` docs (#2515)
typelevel/cats-effect: docs/std/dispatcher.md

![](assets/dispatcher.jpeg)

> `Resource` is a fiber-based [Supervisor](https://typelevel.org/cats-effect/docs/std/supervisor) utility for evaluating effects across an impure boundary. 

Reviewer: You probably meant `Dispatcher` instead of `Resource`?

This is useful when working with reactive interfaces that produce potentially many values (as opposed to one), and for each value, some effect 

> in F must be performed (like inserting it into a queue).

Reviewer: like inserting each value into a queue

Let's say we are integrating with an interface that looks like this:

```scala
abstract class ImpureInterface {
  def onMessage(msg: String): Unit

  def init(): Unit = {
    onMessage("init")
  }

  // ...
}
```

As an user we're supposed to create an implementation of this class where we implement `onMessage`.

Let's say that we have a [`queue`](https://typelevel.org/cats-effect/docs/std/queue) and that we want to add each message received to the queue. We might,
naively, try to do the following

```scala
for {
    queue <- Queue.unbounded[IO, String]
    impureInterface = new ImpureInterface {
      override def onMessage(msg: String): Unit = queue.offer(msg) // This returns an IO, so nothing really happens!
    }
    _ <- IO.delay(impureInterface.init())
    value <- queue.tryTake
  } yield value match {
    case Some(v) => println(s"Value found in queue! $v")
    case None    => println("Value not found in queue :(")
  }
```

This will print "Value not found in queue :(", since our implementation of `onMessage` 

> doesn't really *do* anything. It just creates an `IO`. 

Reviewer: It just creates an `IO` value, which is just a description of an effect, without actually executing it.

Reviewer: Here the `scalac` option `-Ywarn-value-discard` might help hint this problem.

![](assets/dispatcher.jpeg)

> `Dispatcher` is a fiber-based [Supervisor](https://typelevel.org/cats-effect/docs/std/supervisor) utility for evaluating effects across an impure boundary. 

Reviewer: I'm not completely sure we should link `Dispatcher` to `Supervisor` just yet. Unfortunately, `Dispatcher` does not offer the backpressure that `Supervisor` does on scope exit, yet. We have planned improvements to `Dispatcher` coming most likely in the Cats Effect `3.4.x` lineage of releases.

For now, I would prefer if we used the old wording that `Dispatcher` is a functional interface with derivable instances for any effect type conforming to the `Async` typeclass.

Reviewer 2: Huh, I didn't realize that ... what guarantees does `Dispatcher` offer on scope exit then?

Btw this paragraph is taken straight from the current scaladocs actually 😆

Reviewer 2: Yeah, seems like we should fix the scaladocs ...
https://github.com/typelevel/cats-effect/blob/b2276e5a96766fa9df9a86999cc1be246a43bd06/std/shared/src/main/scala/cats/effect/std/Dispatcher.scala#L37-L38

Reviewer: Nothing, there are reported gotchas. The worst thing that happens I guess is that a submitted effect doesn't get a chance to start at all.

Reviewer 2: > The worst thing that happens I guess is that a submitted effect doesn't get a chance to start at all.

I see. So but does it backpressure on _running_ effects? It's just it may close before the queue is empty, hence a submitted effect doesn't start? And does it not start silently, or with a "dispatcher closed" error (I've gotten a few of those in my day :).

Reviewer: I honestly cannot answer this question correctly. 😕

Reviewer: ```suggestion
`Dispatcher` is a fiber-based utility for evaluating effects across an impure boundary. An instance of `Dispatcher` can be derived for any effect type conforming to the `Async` typeclass.
```

Reviewer 2: FWIW, using it for _any_ Async can land you into all sorts of trouble 😛 e.g. see https://github.com/typelevel/cats-effect/issues/2448#issuecomment-959181978.

Reviewer 2: Maybe we should just fess up a bit in the docs and add a caveats section 😅

Reviewer: Sure, I don't mind.

> This is useful when working with reactive interfaces that produce potentially many values (as opposed to one), and for each value, some effect in F must be performed 

Reviewer 2: ```suggestion
This is useful when working with reactive interfaces that produce potentially many values (as opposed to one), and for each value, some effect in `F` must be performed 
```

(like inserting each value into a queue).

Let's say we are integrating with an interface that looks like this:

```scala
abstract class ImpureInterface {
  def onMessage(msg: String): Unit

  def init(): Unit = {
    onMessage("init")
  }

  // ...
}
```

> As an user we're supposed to create an implementation of this class where we implement `onMessage`.

Reviewer 2: ```suggestion
As a user we're supposed to create an implementation of this class where we implement `onMessage`.
```

Let's say that we have a [`queue`](https://typelevel.org/cats-effect/docs/std/queue) and that we want to add each message received to the queue. We might,
naively, try to do the following

```scala
for {
    queue <- Queue.unbounded[IO, String]
    impureInterface = new ImpureInterface {
      override def onMessage(msg: String): Unit = queue.offer(msg) // This returns an IO, so nothing really happens!
    }
    _ <- IO.delay(impureInterface.init())
    value <- queue.tryTake
  } yield value match {
    case Some(v) => println(s"Value found in queue! $v")
    case None    => println("Value not found in queue :(")
  }
```

This will print "Value not found in queue :(", since our implementation of `onMessage` 
doesn't really *do* anything. It just creates an `IO` value which is just a description of an effect,

> without actually executing it. (Here the `scalac` option `-Ywarn-value-discard` might help hint this problem.)

Reviewer 2: ```suggestion
without actually executing it. (Here the `scalac` option `-Ywarn-value-discard` might help hint at this problem.)
```

It is in these cases that `Dispatcher` comes in handy. Here's how it could be used:

```scala

> Dispatcher[IO].use { dispatcher =>
>     for {
>       queue <- Queue.unbounded[IO, String]
>       impureInterface <-
>         IO.delay {
>           new ImpureInterface {
>             override def onMessage(msg: String): Unit =
>               dispatcher.unsafeRunSync(queue.offer(msg))
>           }
>         }
>       _ <- IO.delay(impureInterface.init())
>       value <- queue.tryTake
>     } yield value match {
>       case Some(v) => println(s"Value found in queue! $v")
>       case None    => println("Value not found in queue :(")
>     }

Reviewer 2: ```suggestion
Dispatcher[IO].use { dispatcher =>
  for {
    queue <- Queue.unbounded[IO, String]
    impureInterface <-
      IO.delay {
        new ImpureInterface {
          override def onMessage(msg: String): Unit =
            dispatcher.unsafeRunSync(queue.offer(msg))
        }
      }
    _ <- IO.delay(impureInterface.init())
    value <- queue.tryTake
  } yield value match {
    case Some(v) => println(s"Value found in queue! $v")
    case None    => println("Value not found in queue :(")
  }
}
```

Reviewer: Is this safe to commit?

Reviewer 2: Should be I think, adds a closing brace and fixes the indent.

> ```scala
> Dispatcher[IO].use { dispatcher =>
>     for {
>       queue <- Queue.unbounded[IO, String]
>       impureInterface <-
>         IO.delay {
>           new ImpureInterface {
>             override def onMessage(msg: String): Unit =
>               dispatcher.unsafeRunSync(queue.offer(msg))
>           }
>         }
>       _ <- IO.delay(impureInterface.init())
>       value <- queue.tryTake
>     } yield value match {
>       case Some(v) => println(s"Value found in queue! $v")
>       case None    => println("Value not found in queue :(")
>     }
> ```

Reviewer 2: ```suggestion
```scala
Dispatcher[IO].use { dispatcher =>
  for {
    queue <- Queue.unbounded[IO, String]
    impureInterface <-
      IO.delay {
        new ImpureInterface {
          override def onMessage(msg: String): Unit =
            dispatcher.unsafeRunSync(queue.offer(msg))
        }
      }
    _ <- IO.delay(impureInterface.init())
    value <- queue.tryTake
  } yield value match {
    case Some(v) => println(s"Value found in queue! $v")
    case None    => println("Value not found in queue :(")
  }
}
```
```

Reviewer 2: Woops, it seems it loses the ` ``` ` in a GH comments suggestion.

}
```

> An instance of `Dispatcher` is very cheap - it is encouraged to instantiate it 
> where necessary rather than wiring a single instance throughout an application.

Reviewer 2: ```suggestion
Creating an instance of `Dispatcher` is very cheap - you are encouraged to instantiate it 
where necessary rather than wiring a single instance throughout an application.
```

> ## Cats-Effect 2
> 
> Users of cats effect 2 may be familiar with the `Effect` and `ConcurrentEffect`

Reviewer 2: Sorry for all the tiny nits!

```suggestion
## Cats Effect 2

Users of Cats Effect 2 may be familiar with the `Effect` and `ConcurrentEffect`
```

> typeclasses. These have been removed as they contrained implementations of the

Reviewer 2: ```suggestion
typeclasses. These have been removed as they constrained implementations of the
```

```

An instance of `Dispatcher` is very cheap - it allocates a single fiber so it

> is encouraged to instantiate it where necessary rather than wiring

Reviewer: Let's remove the `it allocates a single fiber` part of the sentence, since this is no longer true. Let's go with it's encouraged to instantiate it...

Reviewer 2: I think this also needs to be updated in the scaladocs?

a single instance throughout an application.

## Cats-Effect 2

Users of cats effect 2 may be familiar with the `Effect` and `ConcurrentEffect`
typeclasses. These have been removed as they contrained implementations of the
typeclasses too much by forcing them to be embeddable in `IO` via `def
toIO[A](fa: F[A]): IO[A]`. However, these typeclasses also had a valid use-case
for unsafe running of effects to interface with impure APIs (`Future`, `NIO`,

> etc).

Reviewer: The last sentence sounds like this is no longer true. Does `Dispatcher` not cover these cases?

Author: This part was copied from the current docs. It made more sense in that context since it was followed by "`Dispatcher` now addresses this use-case.". I'm going to add that sentence.
